WooCommerce/Classes/ViewRelated/Orders/Order Creation/Taxes/NewTaxRateSelectorViewModel.swift

```diff
@@ -1,24 +1,164 @@
+    let onLoadTrigger: PassthroughSubject<Void, Never> = PassthroughSubject()
+
+    /// View models for placeholder rows. Strings are visible to the user as it is shimmering (loading)
+    let placeholderRowViewModels: [TaxRateViewModel] = [Int64](0..<3).map { index in
+        TaxRateViewModel(id: index, name: "placeholder", rate: "10%")
+    }
+
+    init(siteID: Int64,
+         wpAdminTaxSettingsURLProvider: WPAdminTaxSettingsURLProviderProtocol = WPAdminTaxSettingsURLProvider(),
+         stores: StoresManager = ServiceLocator.stores,
+         storageManager: StorageManagerType = ServiceLocator.storageManager) {
+        self.siteID = siteID
         self.wpAdminTaxSettingsURLProvider = wpAdminTaxSettingsURLProvider
+        self.stores = stores
+        self.storageManager = storageManager
+        self.paginationTracker = PaginationTracker(pageFirstIndex: 1, pageSize: 25)
+
+        configureResultsController()
+        configurePaginationTracker()
+        configureFirstPageLoad()
     }
 
     /// WPAdmin URL to navigate user to edit the tax settings
     var wpAdminTaxSettingsURL: URL? {
         wpAdminTaxSettingsURLProvider.provideWpAdminTaxSettingsURL()
     }
+
+    private lazy var resultsController: ResultsController<StorageTaxRate> = {
+        let predicate = NSPredicate(format: "siteID == %lld", siteID)
+        let sortDescriptorByID = NSSortDescriptor(keyPath: \StorageTaxRate.id, ascending: true)
+        let resultsController = ResultsController<StorageTaxRate>(storageManager: storageManager,
+                                                                    matching: predicate,
+                                                                    sortedBy: [ sortDescriptorByID])
+        return resultsController
+    }()
+
+    func onLoadNextPageAction() {
+        paginationTracker.ensureNextPageIsSynced()
+    }
+
+    func onRefreshAction(completion: @escaping () -> Void) {
+        paginationTracker.resync(reason: nil) {
+            completion()
+        }
+    }
+}
+
+extension NewTaxRateSelectorViewModel: PaginationTrackerDelegate {
+    func sync(pageNumber: Int, pageSize: Int, reason: String?, onCompletion: SyncCompletion?) {
+        transitionToSyncingState()
+
+        let action = TaxAction.retrieveTaxRates(siteID: siteID, pageNumber: pageNumber, pageSize: pageSize) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(let results):
+                let hasNextPage = results.count == pageSize
+                onCompletion?(.success(hasNextPage))
+
+                if self.taxRateViewModels.isEmpty {
+                    self.syncState = .empty
+                } else if results.isEmpty {
+                    // We had results previously, but we didn't have any on this page request. Transition to results to stop the syncing visuals.
+                    self.transitionToResultsUpdatedState()
+                }
```

Comment:
Minor question: would it be better if this was done elsewhere, e.g. in the `onCompletion` handler? Reading it, I'm guessing that the completion handler or something else will have set the `taxRateViewModels = []` if that's appropriate from the results, but I can't tell that for sure while reading the code.

I think we're probably forced into doing it here by the design of `PaginationTrackerDelegate`, but just in case we're not, I thought it worth asking.

Reply:
That's a good point right there. I also tried to find another place to do this, but the only place is when we're listening to changes in the results controller. Since these cases don't bring any change in the results controller, we are not notified there.

The completion handler is part of the `PaginationTracker`, which doesn't handle the view model state.